hs_restclient/__init__.py

```diff
@@ -1 +1,666 @@
+
+        Arguments in tuple passed to constructor must be: (url, status_code, params).
+        url and status_code are of type string, while the optional params argument
+        should be a dict.
+    """
+    def __init__(self, args):
+        super(HydroShareHTTPException, self).__init__(args)
+        self.url = args[0]
+        self.method = args[1]
+        self.status_code = args[2]
+        if len(args) >= 4:
+            self.params = args[3]
+        else:
+            self.params = None
+
+    def __str__(self):
+        msg = "Received status {status_code} {status_msg} when accessing {url} " + \
+              "with method {method} and params {params}."
+        return msg.format(status_code=self.status_code,
+                          status_msg=http_responses[self.status_code],
+                          url=self.url,
+                          method=self.method,
+                          params=self.params)
+
+    def __unicode__(self):
+        return self.__str__()
+
+
+class HydroShare(object):
+    """
+        Construct HydroShare object for querying HydroShare's REST API
+
+        :param hostname: Hostname of the HydroShare server to query
+        :param auth: Concrete instance of AbstractHydroShareAuth (i.e. HydroShareAuthBasic)
+        :param use_https: Boolean, if True, HTTPS will be used
+        :param port: Integer representing the TCP port on which to connect
+            to the HydroShare server
+
+        :raises: HydroShareException if auth is not a known authentication type.
+    """
+
+    _URL_PROTO_WITHOUT_PORT = "{scheme}://{hostname}/hsapi"
+    _URL_PROTO_WITH_PORT = "{scheme}://{hostname}:{port}/hsapi"
+
+    def __init__(self, hostname='www.hydroshare.org', auth=None,
+                 use_https=False, port=None):
+        self.hostname = hostname
+
+        self.session = None
+        self.auth = None
+        if auth:
+            if isinstance(auth, HydroShareAuthBasic):
+                # HTTP basic authentication
+                self.auth = (auth.username, auth.password)
+            else:
+                raise HydroShareException("Unsupported authentication type '{0}'.".format(str(type(auth))))
+
+        if use_https:
+            self.scheme = 'https'
+        else:
+            self.scheme = 'http'
+        if port:
+            self.port = int(port)
+            if self.port < 0 or self.port > 65535:
+                raise HydroShareException("Port number {0} is illegal.".format(self.port))
+            self.url_base = self._URL_PROTO_WITH_PORT.format(scheme=self.scheme,
+                                                             hostname=self.hostname,
+                                                             port=self.port)
+        else:
+            self.url_base = self._URL_PROTO_WITHOUT_PORT.format(scheme=self.scheme,
+                                                                hostname=self.hostname)
```

Comment:
The `urlparse` library can also construct URLs for you and handle all the optional bits like ports.